### ISSUE TITLE
support optional tabs in source code

### ISSUE THREAD

**Issue by mrjbq7** [factor/factor#2143]
Can't find the original request for this, but one idea might be to change the ``lexer`` to optionally support tabs, something like this:

```diff
diff --git a/core/lexer/lexer.factor b/core/lexer/lexer.factor
index 7dc514b189..7677128fe6 100644
--- a/core/lexer/lexer.factor
+++ b/core/lexer/lexer.factor
@@ -11,7 +11,8 @@ TUPLE: lexer
 { line-text string }
 { line-length fixnum }
 { column fixnum }
-{ parsing-words vector } ;
+{ parsing-words vector }
+{ forbid-tabs? initial: t } ;
 
 TUPLE: lexer-parsing-word word line line-text column ;
 
@@ -50,7 +51,17 @@ ERROR: not-a-lexer object ;
 ERROR: unexpected want got ;
 
 : forbid-tab ( c -- c )
-    [ CHAR: \t eq? [ "[space]" "[tab]" unexpected ] when ] keep ; inline
+    dup CHAR: \t eq? [
+        lexer get forbid-tabs?>> [
+            "input contains tabs"
+            { { "Ignore Tabs" t } { "Continue" f } }
+            throw-restarts [
+                f lexer get forbid-tabs?<< drop CHAR: \s
+            ] [
+                "[space]" "[tab]" unexpected
+            ] if
+        ] [ drop CHAR: \s ] if
+    ] when ; inline
 
 : skip ( i seq ? -- n )
     over length [

```

**Comment by mrjbq7** [comment]
It's a slippery slope before you support all possible unicode whitespace... plus these are kinda hidden to the developer.

**Comment by LeoMehraban** [comment]
why would it be bad to support all unicode whitespaces? is it ambiguous what is considered a whitespace or something?

**Comment by mrjbq7** [comment]
Performance, simplicity, requiring Unicode on all platforms, more difficult to port, some debate on the extent of Unicode whitespace values, etc.

**Comment by LeoMehraban** [comment]
according to [wikipedia](https://en.wikipedia.org/wiki/Whitespace_character#Programming_languages):

> Most languages only recognize whitespace characters that have an ASCII code. They disallow most or all of the Unicode codes listed above. The [C language](https://en.wikipedia.org/wiki/C_language) defines whitespace characters to be "space, horizontal tab, new-line, vertical tab, and form-feed"

whitespace characters that are in ASCII seems like a natural place to draw the line. that's the best we can do without needing to worry about unicode support. 

in terms of performance, you're the expert (at least compared to me), and so I won't comment. As for simplicity, I can see both sides of the argument

**Comment by mrjbq7** [comment]
Yep, always tradeoffs. Unclear to me if we gain any or much benefit with
widening the accepted values beyond spaces and new lines.
